python/ray/data/_internal/logical/rules/operator_fusion.py

```diff
@@ -139,52 +170,65 @@ def _can_fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator) -> bool:
-            and isinstance(up_logical_op, AbstractUDFMap)
-            and get_compute(up_logical_op._compute)
-            != get_compute(down_logical_op._compute)
+        if isinstance(down_logical_op, AbstractUDFMap) and isinstance(
+            up_logical_op, AbstractUDFMap
         ):
-            return False
-
-        # Fusing callable classes is only supported if they are the same function AND
-        # their construction arguments are the same. Note the Write can be compatbile
-        # with any UDF as Write itself doesn't have UDF.
-        # TODO(Clark): Support multiple callable classes instantiating in the same actor
-        # worker.
-        if (
-            isinstance(down_logical_op, AbstractUDFMap)
-            and isinstance(down_logical_op._fn, CallableClass)
-            and isinstance(up_logical_op, AbstractUDFMap)
-            and isinstance(up_logical_op._fn, CallableClass)
-            and (
-                up_logical_op._fn != down_logical_op._fn
-                or (
-                    up_logical_op._fn_constructor_args
-                    != down_logical_op._fn_constructor_args
-                    or up_logical_op._fn_constructor_kwargs
-                    != down_logical_op._fn_constructor_kwargs
+            # Allow fusing tasks->actors if the resources are compatible (read->map),
+            # but not the other way around. The latter (downstream op) will be used as
+            # the compute if fused.
+            if is_task_compute(down_logical_op._compute) and get_compute(
+                up_logical_op._compute
+            ) != get_compute(down_logical_op._compute):
+                return False
+
+            # Fusing callable classes is only supported if they are the same function
+            # AND their construction arguments are the same. Note the Write can be
+            # compatible  with any UDF as Write itself doesn't have UDF.
+            # TODO(Clark): Support multiple callable classes instantiating
+            # in the same actor worker.
+            if (
+                isinstance(down_logical_op._fn, CallableClass)
+                and isinstance(up_logical_op._fn, CallableClass)
+                and (
+                    up_logical_op._fn != down_logical_op._fn
+                    or (
+                        up_logical_op._fn_constructor_args
+                        != down_logical_op._fn_constructor_args
+                        or up_logical_op._fn_constructor_kwargs
+                        != down_logical_op._fn_constructor_kwargs
+                    )
                 )
-            )
-        ):
-            return False
+            ):
+                return False
 
         # Only fuse if the ops' remote arguments are compatible.
-        if not _are_remote_args_compatible(
+        if (
+            hasattr(up_logical_op, "_ray_remote_args")
+            and hasattr(down_logical_op, "_ray_remote_args")
+        ) and not _are_remote_args_compatible(
```

Comment:
Not a blocker for this PR. The code logic in this file is becoming too complex. We need to refactor it. Ideally, have different functions/subclasses handling different fusion cases.

Reply:
Agreed, in a future PR, we should separate each operator pair fusion into its own rule.